ddr/lib/ddr-ir/Macro.cpp

```diff
@@ -26,56 +26,563 @@
 
 #include "ddr/std/sstream.hpp"
 #include <string.h>
+#include <algorithm>
+#include <vector>
 
 using std::string;
 using std::stringstream;
 
-DDR_RC
-Macro::getNumeric(long long *ret)
+/* represents the different precedence levels of binary operators that are valid for the parser in ascending order */
+enum binExpr_type {
+	LOGICAL_OR, LOGICAL_AND, BITWISE_OR, BITWISE_XOR, BITWISE_AND, EQUALITY, INEQUALITY, BITSHIFT, ADDITION, MULTIPLICATION
+};
+
+class MacroParser
 {
-	DDR_RC rc = DDR_RC_OK;
-	string value = getValue();
+/* data members */
+private:
+	size_t _index; /* keeps track of where the parser is in the expression */
+	std::string _expression; /* string representation of the macro value */
 
-	if (value.length() > 0) {
-		/* If the pre-processed macro contains no brackets, read it as a number. */
-		if (string::npos == value.find('(')) {
-			stringstream ss;
-			ss << value;
-			long long valueNumeric = 0;
-			ss >> valueNumeric;
-			if (ss.fail()) {
-				rc = DDR_RC_ERROR;
-			} else if (NULL != ret) {
-				*ret = valueNumeric;
-			}
-		} else {
-			/* For macros containing brackets, extract the number. This works for casts
-			 * such as ((int)5).
+/* function members */
+private:
+	bool binaryExpression(int64_t *, binExpr_type);
+	bool unaryExpression(int64_t*);
+	bool parentheses(int64_t*);
+	bool readNumber(int64_t*);
+	void skipWhitespace();
+
+public:
+	MacroParser(std::string expression) : _expression(expression), _index(0)
```

Comment:
The variables are being initialized in the wrong order.  They need to be initialized in declaration order since that is the order they will *actually* be initialized in. 

Could you also move them each onto their own line for clarity?